From 9f04bd5df403508700f8bfa6304c5e1bc456f8ee Mon Sep 17 00:00:00 2001
From: ila <ila.embsys@gmail.com>
Date: Fri, 9 Oct 2020 21:02:34 +0300
Subject: [PATCH] ada: Fix generation of specification for union anonymous
 fields

For anonymous field DECL_NAME does not resolves as true to fall to branch with union processing.

So it moved above of DECL_NAME  condition.
Also the change save the previous semantic to process union if DECL_VIRTUAL_P is false
---
 gcc/c-family/c-ada-spec.c | 28 ++++++++++++++--------------
 1 file changed, 14 insertions(+), 14 deletions(-)

diff --git a/gcc/c-family/c-ada-spec.c b/gcc/c-family/c-ada-spec.c
index 76e736964de..0c50bdcd856 100644
--- a/gcc/c-family/c-ada-spec.c
+++ b/gcc/c-family/c-ada-spec.c
@@ -3210,6 +3210,20 @@ dump_ada_structure (pretty_printer *buffer, tree node, tree type, bool nested,
   /* Print the non-static fields of the structure.  */
   for (tree tmp = TYPE_FIELDS (node); tmp; tmp = TREE_CHAIN (tmp))
     {
+      if (!DECL_VIRTUAL_P (tmp) && is_union)
+        {
+          if (TREE_CHAIN (tmp)
+              && TREE_TYPE (TREE_CHAIN (tmp)) != node
+              && TREE_CODE (TREE_CHAIN (tmp)) != TYPE_DECL)
+          sprintf (buf, "when %d =>", field_num);
+          else
+          sprintf (buf, "when others =>");
+
+          INDENT (spc + INDENT_INCR * 2);
+          pp_string (buffer, buf);
+          pp_newline (buffer);
+        }
+
       /* Add parent field if needed.  */
       if (!DECL_NAME (tmp))
 	{
@@ -3242,20 +3256,6 @@ dump_ada_structure (pretty_printer *buffer, tree node, tree type, bool nested,
 	  /* Skip internal virtual table field.  */
 	  if (!DECL_VIRTUAL_P (tmp))
 	    {
-	      if (is_union)
-		{
-		  if (TREE_CHAIN (tmp)
-		      && TREE_TYPE (TREE_CHAIN (tmp)) != node
-		      && TREE_CODE (TREE_CHAIN (tmp)) != TYPE_DECL)
-		    sprintf (buf, "when %d =>", field_num);
-		  else
-		    sprintf (buf, "when others =>");
-
-		  INDENT (spc + INDENT_INCR * 2);
-		  pp_string (buffer, buf);
-		  pp_newline (buffer);
-		}
-
 	      if (dump_ada_declaration (buffer, tmp, type, field_spc))
 		{
 		  pp_newline (buffer);
-- 
2.28.0.windows.1
